### PR TITLE
Remove old QJITDevice class and references

### PR DESCRIPTION
**Pull request by rauletorresc**
**Context:** QJITDevice has been replaced with QJITDeviceNewAPI.

**Description of the Change:** Remove old QJITDevice class and references. Rename QJITDeviceNewAPI as QJITDevice.

**Benefits:** Updated API.

[sc-67124]

### CODE REVIEW

**Comment by rmoyard** [comment]
Is lightning kokkos using the new device API now? Also what about Qrack? @WrathfulSpatula

**Review by rmoyard — COMMENTED** [review]
Could we call it the `QJITDevice`? I don't think the NewAPI is relevant anymore. We should make clear everywhere in the docs that we only support the newest one.

**Comment by dime10** [comment]
> Is lightning kokkos using the new device API now? Also what about Qrack? @WrathfulSpatula

They do not, but the shim should take care of that right?

File: frontend/catalyst/qfunc.py [hunk 1]
```diff
@@ -129,10 +128,8 @@ def __call__(self, *args, **kwargs):
         device_capabilities = get_device_capabilities(self.device, program_features)
         backend_info = QFunc.extract_backend_info(self.device, device_capabilities)
 
-        if isinstance(self.device, qml.devices.Device):
-            qjit_device = QJITDeviceNewAPI(self.device, device_capabilities, backend_info)
-        else:
-            qjit_device = QJITDevice(self.device, device_capabilities, backend_info)
+        assert isinstance(self.device, qml.devices.Device), "Unsupported device API"
```
Comment (dime10) [inline frontend/catalyst/qfunc.py:131]: This could feasibly be triggered by a user so an exception would be better

Comment (rauletorresc) [reply frontend/catalyst/qfunc.py]: Done

**Comment by rauletorresc** [comment]
> Could we call it the `QJITDevice`? I don't think the NewAPI is relevant anymore. We should make clear everywhere in the docs that we only support the newest one.

Where can I find this in the docs?

**Comment by rauletorresc** [comment]
QJITDeviceNewAPI is QJITDevice now

**Comment by rmoyard** [comment]
Here there is stuff to be updated: `catalyst/doc/dev/custom_devices.rst` it is still mentionning the old/new device API. See at the end

File: frontend/catalyst/qfunc.py (hunk 1 above)
Comment (rauletorresc) [reply frontend/catalyst/qfunc.py]: It seems the proposed exception will hardly be reached because Pennylane has removed the legacy `qml.Device` interface and replaced it with an abstract class called `qml.devices.LegacyDevice`. On top of that, the new `qml.device` converts all Legacy Devices to the new API with the `LegacyDeviceFacade` class. Moreover, a Custom Legacy Device derived from the abstract class would require also a toml specification and a shared object in order to work with Catalyst. We are deleting the documentation section that describes how to use the legacy API.

In short, to test the proposed exception we would require a legacy device with toml spec at least, so we pass the compilation constrains. Is it worth doing that for a case that is very unlikely to happen? 

I would say: let's turn into an assertion in the meantime.

**Comment by rauletorresc** [comment]
> Here there is stuff to be updated: `catalyst/doc/dev/custom_devices.rst` it is still mentionning the old/new device API. See at the end

Updated

File: frontend/catalyst/qfunc.py (hunk 1 above)
Comment (rmoyard) [reply frontend/catalyst/qfunc.py]: Can we just remove it?

Comment (rmoyard) [reply frontend/catalyst/qfunc.py]: From what I understand, we still support the old device API through the facade.

File: frontend/catalyst/third_party/cuda/__init__.py [hunk 2]
```diff
@@ -106,7 +106,7 @@ class BaseCudaInstructionSet(qml.QubitDevice):
     # There are similar lines of code in possibly
     # all other list of operations supported by devices.
     # At the time of writing, this warning is raised
-    # due to similar lines of code in the QJITDevice
+    # due to similar lines of code in the QJITDeviceNewAPI
```
Comment (rmoyard) [inline frontend/catalyst/third_party/cuda/__init__.py]:
```suggestion
    # due to similar lines of code in the QJITDevice
```

File: frontend/catalyst/qfunc.py (hunk 1 above)
Comment (erick-xanadu) [reply frontend/catalyst/qfunc.py]: I agree with Raul.

Comment (rauletorresc) [reply frontend/catalyst/qfunc.py]: But the `LegacyDeviceFacade `type is in fact an instance of `qml.devices.Device`, hence the condition for the exception will never be met:
```
if not isinstance(self.device, qml.devices.Device):
  raise TypeError("Unsupported device API. Please use 'qml.devices.Device' instead.")
```

File: frontend/catalyst/third_party/cuda/__init__.py (hunk 2 above)
Comment (rauletorresc) [reply frontend/catalyst/third_party/cuda/__init__.py]: It was changed already

File: frontend/catalyst/qfunc.py [hunk 3]
```diff
@@ -129,10 +128,11 @@ def __call__(self, *args, **kwargs):
         device_capabilities = get_device_capabilities(self.device, program_features)
         backend_info = QFunc.extract_backend_info(self.device, device_capabilities)
 
-        if isinstance(self.device, qml.devices.Device):
-            qjit_device = QJITDeviceNewAPI(self.device, device_capabilities, backend_info)
-        else:
-            qjit_device = QJITDevice(self.device, device_capabilities, backend_info)
+        assert isinstance(
```
Comment (rmoyard) [inline frontend/catalyst/qfunc.py:131]: I would remove it

**Review by rmoyard — COMMENTED** [review]
Happy to approve after removal of the assert.

File: frontend/catalyst/qfunc.py (hunk 3 above)
Comment (rauletorresc) [reply frontend/catalyst/qfunc.py]: Removed
